<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to mitolina: MITOchondrial LINeage Analysis • mitolina</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to mitolina: MITOchondrial LINeage Analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mitolina</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction to mitolina: MITOchondrial LINeage Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to mitolina: MITOchondrial LINeage Analysis</h1>
                        <h4 class="author">Mikkel Meyer Andersen</h4>
            
            <h4 class="date">02 May, 2018</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>First, the library is loaded:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(mitolina)</code></pre></div>
<p>For reproducibility, the seed for the (pseudo) random number generator is set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)</code></pre></div>
<p>In this vignette, I will use functionality from the <code>dplyr</code> (that loads the pipe operator from the <code>magrittr</code> package). More information about these are available at <a href="https://magrittr.tidyverse.org/" class="uri">https://magrittr.tidyverse.org/</a> and <a href="http://dplyr.tidyverse.org/" class="uri">http://dplyr.tidyverse.org/</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)</code></pre></div>
</div>
<div id="simulating-genealogies" class="section level2">
<h2 class="hasAnchor">
<a href="#simulating-genealogies" class="anchor"></a>Simulating genealogies</h2>
<p>A genealogy is simulated by specifying population sizes for females and males.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sample_mtdna_geneology_varying_size.html">sample_mtdna_geneology_varying_size</a></span>(
  <span class="dt">population_sizes_females =</span> <span class="kw">rep</span>(<span class="dv">10</span>, <span class="dv">10</span>),
  <span class="dt">population_sizes_males =</span> <span class="kw">c</span>(0L, <span class="kw">rep</span>(<span class="dv">10</span>, <span class="dv">9</span>)),
  <span class="dt">progress =</span> <span class="ot">FALSE</span>)
pop &lt;-<span class="st"> </span>sim$population
pop</code></pre></div>
<pre><code>## Population with 178 individuals</code></pre>
<p>Note, that the first male generation must have size 0.</p>
<p>Please refer to the documentation of <code>sample_mtdna_geneology_varying_size</code> (<code><a href="../reference/sample_mtdna_geneology_varying_size.html">?sample_mtdna_geneology_varying_size</a></code>) for details in variability in reproductive success (VRS) <span class="citation">(Andersen and Balding 2017)</span>.</p>
<p>This only sample a genealogy. We also need to build pedigrees:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">peds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/build_pedigrees.html">build_pedigrees</a></span>(pop, <span class="dt">progress =</span> <span class="ot">FALSE</span>)
peds</code></pre></div>
<pre><code>## List of 8 pedigrees (of size 3, 3, 15, 5, 46, 94, ...)</code></pre>
<p>We can then plot one pedigree. We take a pedigree with a male individual in the final generation (<code>end_generation_male_individuals</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">end_gen_peds &lt;-<span class="st"> </span><span class="kw">lapply</span>(sim$end_generation_male_individuals, get_pedigree_from_individual)
pedids &lt;-<span class="st"> </span><span class="kw">sapply</span>(end_gen_peds, get_pedigree_id)
id_tab &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">table</span>(pedids))
ped_id_endgen &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">names</span>(id_tab)[<span class="dv">1</span>]) <span class="co"># smallest</span>
indv_index &lt;-<span class="st"> </span><span class="kw">which</span>(pedids ==<span class="st"> </span>ped_id_endgen)[<span class="dv">1</span>]
ped &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_pedigree_from_individual.html">get_pedigree_from_individual</a></span>(sim$end_generation_male_individuals[[indv_index]])
<span class="kw"><a href="../reference/pedigree_size.html">pedigree_size</a></span>(ped)</code></pre></div>
<pre><code>## [1] 46</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(ped)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>The squares are males and the circles are females.</p>
</div>
<div id="adding-mitogenomes" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-mitogenomes" class="anchor"></a>Adding mitogenomes</h2>
<p>We are now ready to add haplotypes/mitogenomes to the individuals. Let us consider</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sites &lt;-<span class="st"> </span><span class="dv">100</span></code></pre></div>
<p>in the mitogenome for know.</p>
<p>First, we construct a function to create founder haplotypes. Here, they are just random:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">get_founder_mito &lt;-<span class="st"> </span>function() {
  <span class="kw"><a href="http://www.rdocumentation.org/packages/dplyr/topics/sample">sample</a></span>(<span class="kw">c</span>(<span class="ot">FALSE</span>, <span class="ot">TRUE</span>), sites, <span class="dt">replace =</span> <span class="ot">TRUE</span>)
}
x &lt;-<span class="st"> </span><span class="kw">get_founder_mito</span>()
<span class="kw">head</span>(x)</code></pre></div>
<pre><code>## [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(x)</code></pre></div>
<pre><code>## [1] 51</code></pre>
<p>Let us assume the following mutation rates per site:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mu &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="fl">1e-2</span>, sites)</code></pre></div>
<p>Now, we can add mitogenomes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pedigrees_all_populate_haplotypes_custom_founders.html">pedigrees_all_populate_haplotypes_custom_founders</a></span>(
  <span class="dt">pedigrees =</span> peds,
  <span class="dt">mutation_rates =</span> mu,
  <span class="dt">get_founder_haplotype =</span> get_founder_mito, 
  <span class="dt">progress =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>We can then try to plot the same pedigree again, this time with information about the individuals’ mitogenomes (namely, the number of variants in each individuals mitogenome):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(ped, <span class="dt">num_vars =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>Plotting with the <code>igraph</code> plot quickly becomes difficult to customise. Therefore extensions to the <code>tidygraph</code>/<code>ggraph</code> packages is provided:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidygraph)
<span class="kw">library</span>(ggraph)</code></pre></div>
<pre><code>## Loading required package: ggplot2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">g &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/tbl_graph">as_tbl_graph</a></span>(peds) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/activate">activate</a></span>(nodes) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">filter</a></span>(ped_id ==<span class="st"> </span>ped_id_endgen) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">mutate</a></span>(<span class="dt">num_vars =</span> <span class="kw">sapply</span>(haplotype, sum))

p &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/ggraph">ggraph</a></span>(g, <span class="dt">layout =</span> <span class="st">'tree'</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_edge_link">geom_edge_link</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_node_point">geom_node_point</a></span>(<span class="kw">aes</span>(<span class="dt">color =</span> sex, <span class="dt">fill =</span> num_vars, <span class="dt">shape =</span> sex), <span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">stroke =</span> <span class="dv">2</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_node_text">geom_node_text</a></span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">paste0</span>(name, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>, num_vars)), <span class="dt">size =</span> <span class="dv">4</span>, <span class="dt">color =</span> <span class="st">"white"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/theme_graph">theme_graph</a></span>(<span class="dt">base_family =</span> <span class="st">""</span>) +
<span class="st">  </span><span class="kw">scale_shape_manual</span>(<span class="ot">NULL</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"Female"</span> =<span class="st"> </span><span class="dv">21</span>, <span class="st">"Male"</span> =<span class="st"> </span><span class="dv">22</span>)) +
<span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="ot">NULL</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"Female"</span> =<span class="st"> "black"</span>, <span class="st">"Male"</span> =<span class="st"> "red"</span>))
p</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="1152"></p>
<p>Here, the top number is the id and the number below is the number of variants in the individual’s mitogenome. Note, how it is possible to get both a (border) color and fill for <code>pch</code> values 21-25.</p>
<p>We can also extract all haplotypes in this pedigree:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_haplotypes_in_pedigree.html">get_haplotypes_in_pedigree</a></span>(ped)
<span class="kw">length</span>(haps)</code></pre></div>
<pre><code>## [1] 46</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(<span class="kw">sapply</span>(haps, sum))</code></pre></div>
<pre><code>## [1] 51.28261</code></pre>
<p>Or from the last male generation by</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_haplotypes_individuals.html">get_haplotypes_individuals</a></span>(sim$end_generation_male_individuals)
<span class="kw">length</span>(haps)</code></pre></div>
<pre><code>## [1] 10</code></pre>
<p>In the simulation function <code>sample_mtdna_geneology_varying_size</code>, there is a parameter <code>generations_return</code> to decide how many generations are returned in the <code>male_individuals_generations</code> and <code>female_individuals_generations</code> elements. The default is <code>3</code>, meaning that individuals in 3 last generations are returned. So, with the default, the last three generations of males are available in the <code>male_individuals_generations</code> element.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_haplotypes_individuals.html">get_haplotypes_individuals</a></span>(sim$male_individuals_generations)
<span class="kw">length</span>(haps)</code></pre></div>
<pre><code>## [1] 30</code></pre>
</div>
<div id="mtdna-mutation-schemes" class="section level2">
<h2 class="hasAnchor">
<a href="#mtdna-mutation-schemes" class="anchor"></a>mtDNA mutation schemes</h2>
<p>There are several datasets with information about the mitogenome included. There is a help file to each of the datasets which can be found by e.g. <code><a href="../reference/mtdna_partitions.html">?mtdna_partitions</a></code> for the <code>mtdna_partitions</code> dataset.</p>
<p>First, there is information about the positions from PhyloTree <a href="http://www.phylotree.org/resources/rCRS_annotated.htm" class="uri">http://www.phylotree.org/resources/rCRS_annotated.htm</a> about the revised Cambridge Reference Sequence (rCRS) <span class="citation">(Andrews et al. 1999)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_partitions)
mtdna_partitions %&gt;%<span class="st"> </span><span class="kw">print</span>(<span class="dt">n =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>## # A tibble: 16,569 x 5
##   Position rCRS  PartitionPhyloTree PartitionRieux  PartitionSoares
##      &lt;int&gt; &lt;chr&gt; &lt;chr&gt;              &lt;chr&gt;           &lt;chr&gt;          
## 1        1 G     Control_region     [Not included?] CR: Remaining  
## 2        2 A     Control_region     [Not included?] CR: Remaining  
## # ... with 1.657e+04 more rows</code></pre>
<p>There is also mutation information from <span class="citation">(Soares et al. 2009)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_mut_SoaresK8)
mtdna_mut_SoaresK8</code></pre></div>
<pre><code>## # A tibble: 8 x 4
##   PartitionSoares MutRateYearly MutRateSEYearly MutationScheme     
##   &lt;chr&gt;                   &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;              
## 1 PC1 + PC2       0.00000000888  0.000000000755 Soares 2009 (K = 8)
## 2 PC3             0.0000000193   0.000000000755 Soares 2009 (K = 8)
## 3 CR: HVS1        0.000000164    0.000000000755 Soares 2009 (K = 8)
## 4 CR: HVS2        0.000000230    0.000000000755 Soares 2009 (K = 8)
## 5 CR: Remaining   0.0000000155   0.000000000755 Soares 2009 (K = 8)
## 6 rRNA            0.00000000820  0.000000000755 Soares 2009 (K = 8)
## 7 tRNA            0.00000000691  0.000000000755 Soares 2009 (K = 8)
## 8 Others          0.0000000249   0.000000000755 Soares 2009 (K = 8)</code></pre>
<p>From <span class="citation">(Rieux et al. 2014)</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_mut_RieuxK4)
mtdna_mut_RieuxK4</code></pre></div>
<pre><code>## # A tibble: 5 x 6
##   PartitionRieux  MutRateYearly   MutHPDLower  MutHPDUpper MutRateSEYearly
##   &lt;chr&gt;                   &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt;           &lt;dbl&gt;
## 1 PC1 + PC2       0.00000000756 0.00000000571      9.35e-9  0.000000000929
## 2 PC3             0.0000000332  0.0000000257       4.07e-8  0.00000000384 
## 3 HVS1 + HVS2     0.000000314   0.000000226        4.03e-7  0.0000000453  
## 4 rRNA + tRNA     0.0000000101  0.00000000757      1.27e-8  0.00000000130 
## 5 [Not included?] 0             0                  0.       0             
## # ... with 1 more variable: MutationScheme &lt;chr&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_mut_RieuxK1)
mtdna_mut_RieuxK1</code></pre></div>
<pre><code>## # A tibble: 1 x 5
##   MutRateYearly  MutHPDLower  MutHPDUpper MutRateSEYearly MutationScheme  
##           &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;           
## 1  0.0000000214 0.0000000158 0.0000000271   0.00000000287 Rieux 2014 (K =…</code></pre>
<p>And from <span class="citation">(Översti et al. 2017)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_mut_OverstiK4)
mtdna_mut_OverstiK4</code></pre></div>
<pre><code>## # A tibble: 5 x 6
##   PartitionRieux  MutRateYearly  MutHPDLower  MutHPDUpper MutRateSEYearly
##   &lt;chr&gt;                   &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;           &lt;dbl&gt;
## 1 PC1 + PC2        0.000000018  0.0000000117 0.000000024    0.00000000314
## 2 PC3              0.0000000304 0.0000000192 0.0000000421   0.00000000584
## 3 HVS1 + HVS2      0.000000206  0.000000125  0.000000290    0.0000000421 
## 4 rRNA + tRNA      0.0000000158 0.0000000094 0.0000000230   0.00000000347
## 5 [Not included?]  0            0            0              0            
## # ... with 1 more variable: MutationScheme &lt;chr&gt;</code></pre>
<p>All these are assembled info one dataframe:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mtdna_mut_schemes)
mtdna_mut_schemes %&gt;%<span class="st"> </span><span class="kw">print</span>(<span class="dt">n =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>## # A tibble: 66,276 x 4
##   Position MutationScheme      MutRateYearly MutRateSEYearly
##      &lt;int&gt; &lt;chr&gt;                       &lt;dbl&gt;           &lt;dbl&gt;
## 1        1 Soares 2009 (K = 8)  0.0000000155  0.000000000755
## 2        2 Soares 2009 (K = 8)  0.0000000155  0.000000000755
## # ... with 6.627e+04 more rows</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mtdna_mut_schemes %&gt;%<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/dplyr/topics/tally">count</a></span>(MutationScheme)</code></pre></div>
<pre><code>## # A tibble: 4 x 2
##   MutationScheme           n
##   &lt;chr&gt;                &lt;int&gt;
## 1 Oversti 2017 (K = 4) 16569
## 2 Rieux 2014 (K = 1)   16569
## 3 Rieux 2014 (K = 4)   16569
## 4 Soares 2009 (K = 8)  16569</code></pre>
<p>This makes it easier to use the mutation rates from for example <code>mtdna_mut_RieuxK4</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d_mu &lt;-<span class="st"> </span>mtdna_mut_schemes %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">filter</a></span>(MutationScheme ==<span class="st"> "Rieux 2014 (K = 4)"</span>, MutRateYearly &gt;<span class="st"> </span><span class="dv">0</span>)
mu_mean_year &lt;-<span class="st"> </span>d_mu %&gt;%<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">pull</a></span>(MutRateYearly)
mu_mean_gen &lt;-<span class="st"> </span><span class="dv">25</span>*mu_mean_year</code></pre></div>
<p>Now, <code>mu_mean_gen</code> contains a mutation rate per generation (of 25 years) for each of the 16033 sites. The probability of one ore more mutations in a single meiosis is then</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> -<span class="st"> </span><span class="kw">prod</span>(<span class="dv">1</span> -<span class="st"> </span>mu_mean_gen)</code></pre></div>
<pre><code>## [1] 0.01083419</code></pre>
<p>Note, that to include variability in the mutation rates, it is probably best to sample a rate per region (in <code>mtdna_mut_RieuxK4</code>), and then distribute that to all sites in the partition.</p>
<p>These rates are easily used to populate mitogenomes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">get_founder_mito &lt;-<span class="st"> </span>function() {
  <span class="kw"><a href="http://www.rdocumentation.org/packages/dplyr/topics/sample">sample</a></span>(<span class="kw">c</span>(<span class="ot">FALSE</span>, <span class="ot">TRUE</span>), <span class="kw">length</span>(mu_mean_gen), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
}

<span class="kw"><a href="../reference/pedigrees_all_populate_haplotypes_custom_founders.html">pedigrees_all_populate_haplotypes_custom_founders</a></span>(
  <span class="dt">pedigrees =</span> peds,
  <span class="dt">mutation_rates =</span> mu_mean_gen,
  <span class="dt">get_founder_haplotype =</span> get_founder_mito, 
  <span class="dt">progress =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Number of variants in the last three male generations (not just in the one pedigree, but in entire population):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_haplotypes_individuals.html">get_haplotypes_individuals</a></span>(sim$male_individuals_generations)
<span class="kw">table</span>(<span class="kw">sapply</span>(haps, sum))</code></pre></div>
<pre><code>## 
## 7944 7945 8010 
##   25    1    4</code></pre>
<p>Number of variants in the last male generation (not just in the one pedigree, but in entire population):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_haplotypes_individuals.html">get_haplotypes_individuals</a></span>(sim$end_generation_male_individuals)
<span class="kw">table</span>(<span class="kw">sapply</span>(haps, sum))</code></pre></div>
<pre><code>## 
## 7944 8010 
##    9    1</code></pre>
<p>And the same pedigree with new mitogenomes, now where number of variants is a factor to get a discrete fill colour scale:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">g &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/tbl_graph">as_tbl_graph</a></span>(peds) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/activate">activate</a></span>(nodes) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">filter</a></span>(ped_id ==<span class="st"> </span>ped_id_endgen) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidygraph/topics/reexports">mutate</a></span>(<span class="dt">num_vars =</span> <span class="kw">sapply</span>(haplotype, sum))

p &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/ggraph">ggraph</a></span>(g, <span class="dt">layout =</span> <span class="st">'tree'</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_edge_link">geom_edge_link</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_node_point">geom_node_point</a></span>(<span class="kw">aes</span>(<span class="dt">color =</span> sex, <span class="dt">fill =</span> <span class="kw">factor</span>(num_vars), <span class="dt">shape =</span> sex), <span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">stroke =</span> <span class="dv">2</span>, <span class="dt">show.legend =</span> <span class="ot">FALSE</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/geom_node_text">geom_node_text</a></span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">paste0</span>(name, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>, num_vars)), <span class="dt">size =</span> <span class="dv">4</span>, <span class="dt">color =</span> <span class="st">"white"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggraph/topics/theme_graph">theme_graph</a></span>(<span class="dt">base_family =</span> <span class="st">""</span>) +
<span class="st">  </span><span class="kw">scale_shape_manual</span>(<span class="ot">NULL</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"Female"</span> =<span class="st"> </span><span class="dv">21</span>, <span class="st">"Male"</span> =<span class="st"> </span><span class="dv">22</span>)) +
<span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="ot">NULL</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"Female"</span> =<span class="st"> "black"</span>, <span class="st">"Male"</span> =<span class="st"> "red"</span>))
p</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-28-1.png" width="1152"></p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-AndersenPLOSGEN2017">
<p>Andersen, MM, and DJ Balding. 2017. “How convincing is a matching Y-chromosome profile?” <em>PLOS Genetics</em> 13 (11): e1007028.</p>
</div>
<div id="ref-rCRS">
<p>Andrews, Richard M, Iwona Kubacka, Patrick F Chinnery, Robert N Lightowlers, Douglass M Turnbull, and Neil Howell. 1999. “Reanalysis and revision of the Cambridge reference sequence for human mitochondrial DNA.” <em>Nature Genetics</em> 23.</p>
</div>
<div id="ref-Oversti2017">
<p>Översti, Sanni, Päivi Onkamo, Monika Stoljarova, Bruce Budowle, Antti Sajantila, and Jukka U. Palo. 2017. “Identification and analysis of mtDNA genomes attributed to Finns reveal long-stagnant demographic trends obscured in the total diversity.” <em>Scientific Reports</em> 7.</p>
</div>
<div id="ref-Rieux2014">
<p>Rieux, Adrien, Anders Eriksson, Mingkun Li, Benjamin Sobkowiak, Lucy A. Weinert, Vera Warmuth, Andres Ruiz-Linares, Andrea Manica, and François Balloux. 2014. “Improved Calibration of the Human Mitochondrial Clock Using Ancient Genomes.” <em>Molecular Biology and Evolution</em> 31 (10): 2780–92.</p>
</div>
<div id="ref-Soares2009">
<p>Soares, Pedro, Luca Ermini, Noel Thomson, Maru Mormina, Teresa Rito, Arne Röhl, Antonio Salas, Stephen Oppenheimer, Vincent Macaulay, and Martin B Richards. 2009. “Correcting for Purifying Selection: An Improved Human Mitochondrial Molecular Clock.” <em>American Journal of Human Genetics</em> 84: 740–59.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#simulating-genealogies">Simulating genealogies</a></li>
      <li><a href="#adding-mitogenomes">Adding mitogenomes</a></li>
      <li><a href="#mtdna-mutation-schemes">mtDNA mutation schemes</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Mikkel Meyer Andersen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
